experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -71,6 +83,187 @@ LogicalResult HandshakeSpeculationPass::placeUnits(Value ctrlSignal) {
+
+    // Follow the two branch results with a different control signal
+    for (int i = 0; i <= 1; ++i) {
+      for (Operation *succOp : currOp->getResult(i).getUsers()) {
+        Value ctrl = branchCond->getResult(i);
+        bool routed = routeBranchControlTraversal(visited, ctrl, succOp);
+        found_commit = found_commit || routed;
+      }
+    }
+
+    if (!found_commit) {
+      // Remove unused branch signal
+      branchCond->erase();
+      branchDisc->erase();
+    }
+
+  } else {
+    // Continue Traversal
+    for (Value res : currOp->getResults()) {
+      for (Operation *succOp : res.getUsers()) {
+        bool routed = routeBranchControlTraversal(visited, ctrlSignal, succOp);
+        found_commit = found_commit || routed;
+      }
+    }
+  }
+  return found_commit;
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceCommits() {
+  // Place commits and connect to the Speculator Commit Control Signal
+  Value commitCtrl = specOp.getCommitCtrl();
+  if (failed(placeUnits<handshake::SpecCommitOp>(commitCtrl)))
+    return failure();
+
+  // Create visited set
+  llvm::DenseSet<Operation *> visited;
+  visited.insert(specOp.getDataOut().getDefiningOp());
+
+  // Start traversal at the speculator output
+  for (Operation *succOp : specOp.getDataOut().getUsers()) {
+    routeBranchControlTraversal(visited, commitCtrl, succOp);
+  }
+
+  return success();
+}
+
+static LogicalResult findControlBranch(ModuleOp modOp, unsigned bb,
+                                       ConditionalBranchOp &cbranchOp) {
+  handshake::FuncOp funcOp = *modOp.getOps<handshake::FuncOp>().begin();
+  assert(funcOp && "funcOp not found");
+  auto handshakeBlocks = getLogicBBs(funcOp);
+
+  handshake::ConditionalBranchOp controlBranch;
+
+  // Iterate all Ops in the BB where we want to have Speculation
+  for (auto blockOp : handshakeBlocks.blocks.lookup(bb)) {
+    if ((controlBranch = dyn_cast<handshake::ConditionalBranchOp>(blockOp))) {
+      // SaveCommits should only be placed in BBs where there is a Backedge
+      // If this function is called, a backedge should exist
+      for (Value result : blockOp->getResults()) {
+        for (Operation *user : result.getUsers()) {
+          if (isBackedge(result, user)) {
+            cbranchOp = controlBranch;
+            return success();
+          }
+        }
+      }
+    }
+  }
+  llvm::errs() << "Could not find backedge within the Speculator's BB\n";
+  return failure();
+}
+
+LogicalResult HandshakeSpeculationPass::prepareAndPlaceSaveCommits() {
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+
+  // Don't do anything if there are no SaveCommits to place
+  if (placements.getPlacements<handshake::SpecSaveCommitOp>().size() == 0)
+    return success();
+
+  // The save commits are a result of a control branch being in the BB
+  // The control path for the SC needs to replicate the branch
+  unsigned bb = getLogicBB(specOp.getDataIn().getDefiningOp()).value();
```

Comment:
Same as below: surely getDataIn accesses the object you want? is this a type issue, or do we actually need this circular getChild getParent getChild calling?

Reply:
Fixed as stated in a comment above.